metabase/metabase
# Preserve metadata changes for datasets (via editor)

Note: this PR contains both FE and BE changes

This PR allows changing field metadata on a data model/dataset via the editor. The editor modifies queries `results_metadata` locally and once the "Save Changes" button is clicked, they're updated via `PUT /api/card/:id`

### To Verify

I really suggest watching the demo video below before testing the changes to get a better idea of how the editor works and which behaviors are worth testing

For both GUI and native queries:

1. Turn a question into a data model (prev. dataset)
2. Click the data model title to open the details sidebar
3. See the metadata completeness indicator next to the "Customize metadata" button, it should be below 100% (the way it works is described in detail in #19194)
4. Click "Customize metadata"
5. Try editing fields metadata (display name, description, semantic type, formatting, etc.)
6. Optional: try mixing editing the query and the metadata simultaneously
7. Click the "Save changes"  button at the top right at some point
8. Ensure the changes are persisted
9. Try creating a question based on the data model and make sure metadata overwrites work and remain visible (e.g. if you set a field to be an FK, the notebook editor should be able to automatically set a join condition by FK metadata)

**Extra flows**

1. Try to do a few data model edits (change metadata > Save > change metadata > Save > change smth else > Save, etc.) and try reverting to a particular revision, metadata changes should be kept
2. For GUI data models: play around with custom columns metadata editing. Create a few columns and see if their metadata is handled properly

### Demo

https://user-images.githubusercontent.com/17258145/149556601-f50d2b16-4b8d-4f6a-bcd0-c80705122007.mov
